legacy/nosql/mongo/mongo.go

```diff
@@ -396,51 +445,104 @@ func (q *Query) Limit(n int) nosql.Query {
 	q.limit = n
 	return q
 }
-func (q *Query) build() *mgo.Query {
-	var m interface{}
+func (q *Query) build() (*mongo.Cursor, error) {
+	var m interface{} = bson.D{}
 	if q.query != nil {
 		m = q.query
 	}
-	qu := q.c.c.Find(m)
+	findOptions := options.Find()
 	if q.limit > 0 {
-		qu = qu.Limit(q.limit)
+		findOptions.SetLimit(int64(q.limit))
 	}
-	return qu
+
+	qu, err := q.c.c.Find(context.TODO(), m, findOptions)
+
+	return qu, err
 }
 func (q *Query) Count(ctx context.Context) (int64, error) {
-	n, err := q.build().Count()
-	return int64(n), err
+
+	var m interface{} = bson.D{}
+	if q.query != nil {
+		m = q.query
+	}
+	countOptions := options.Count()
+
+	if q.limit > 0 {
+		countOptions.SetLimit(int64(q.limit))
+	}
+
+	count, err := q.c.c.CountDocuments(ctx, m, countOptions)
+
+	if err != nil {
+		return 0, err
+	}
+
+	return count, nil
 }
 func (q *Query) One(ctx context.Context) (nosql.Document, error) {
-	var m bson.M
-	err := q.build().One(&m)
-	if err == mgo.ErrNotFound {
-		return nil, nosql.ErrNotFound
-	} else if err != nil {
+	var m = &bson.M{}
+	cursor, err := q.build()
+
+	if err != nil {
 		return nil, err
 	}
-	return q.c.convDoc(m), nil
+	defer cursor.Close(ctx)
+
+	if cursor.Next(ctx) {
+		if err := cursor.Decode(m); err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, nosql.ErrNotFound
+	}
+	return q.c.convDoc(*m), nil
+
 }
 func (q *Query) Iterate() nosql.DocIterator {
-	it := q.build().Iter()
+	it, err := q.build()
+
+	if err != nil {
+
+		return &Iterator{it: it, err: err, c: q.c}
+	}
 	return &Iterator{it: it, c: q.c}
 }
 
 type Iterator struct {
 	c   *collection
-	it  *mgo.Iter
+	err error
+	it  *mongo.Cursor
 	res bson.M
 }
 
 func (it *Iterator) Next(ctx context.Context) bool {
-	it.res = make(bson.M)
-	return it.it.Next(&it.res)
+	elem := make(bson.M)
+
+	if it.it != nil {
```

Comment:
```suggestion
	if it.it != nil {
		return false
	}
	elem := make(bson.M)
	// the rest of the code
```

Reply:
updated, this one I think your suggested change you meant it.it == nil { return false}